gobblin-core/src/main/java/gobblin/writer/http/AbstractAsyncDataWriter.java

```diff
@@ -0,0 +1,179 @@
+   *
+   * @param buffer The buffer which contains a collection of records
+   * @return the number of records are successfully dispatched or -1 if dispatch failed
+   */
+  protected abstract int dispatch(Queue<BufferedRecord<D>> buffer) throws Throwable;
+
+  private void put(BufferedRecord<D> record) {
+    if (isClosing) {
+      RuntimeException e = new RuntimeException("Attempt to write data when writer is closing");
+      LOG.error("Invalid write", e);
+      throw e;
+    }
+
+    lock.lock();
+    while (buffer.size() == capacity) {
+      try {
+        notFull.await();
+      } catch (InterruptedException e) {
+        LOG.info("Waiting for queue not full interrupted", e);
+      }
+    }
+
+    buffer.add(record);
+    notEmpty.signalAll();
+    lock.unlock();
+  }
+
+  @Override
+  public Future<WriteResponse> write(D record, @Nullable WriteCallback callback) {
+    if (service.isShutdown()) {
+      RuntimeException e = new RuntimeException("Attempt to write a record after close.");
+      LOG.error("Invalid write", e);
+      throw e;
+    }
+    BufferedRecord<D> item = new BufferedRecord<>(record, callback);
+    put(item);
+    return null;
+  }
+
+  @Override
+  public void close()
+      throws IOException {
+    isClosing = true;
+
+    // Invoke RecordProcessor in case it sleeps
+    lock.lock();
+    notEmpty.signalAll();
+    lock.unlock();
+
+    try {
+      recordProcessorFuture.get();
+    } catch (InterruptedException | ExecutionException e) {
+      LOG.info("Waiting for record processor returns an exception", e);
+    }
+    service.shutdown();
+  }
+
+  @Override
+  public void flush()
+      throws IOException {
+    lock.lock();
+    while (buffer.size() != 0) {
+      try {
+        isEmpty.await();
+      } catch (InterruptedException e) {
+        LOG.info("Waiting for buffer empty interrupted", e);
+      }
+    }
+    lock.unlock();
+  }
+
+  private class RecordProcessor implements Runnable {
+    public void run() {
+      LOG.info("Start processing records in the blocking queue");
+      /*
+       * A main loop to process records
+       */
+      while (true) {
+        lock.lock();
+        // Waiting for records to dispatch
+        while (buffer.size() == 0) {
+          if (isClosing) {
+            lock.unlock();
+            // Safely return after processing all records
+            return;
+          }
+          try {
+            notEmpty.await();
+          } catch (InterruptedException e) {
+            LOG.info("Waiting for queue not empty interrupted", e);
+          }
+        }
+
+        // Dispatch records
+        try {
+          if (dispatch(buffer) < 0) {
+            throw new IOException("Failed on dispatching");
+          }
+        } catch (Throwable throwable) {
+          LOG.error("RecordProcessor breaks on exception", throwable);
+          throw new RuntimeException("Exception on dispatching", throwable);
```

Comment:
Yes. Since dispatch may have already retried, all subsequent process would fail for the same reason, meaning the processor is in a bad state and there is no point to keep it alive.